python-checks/src/main/java/org/sonar/python/checks/MembershipTestSupportCheck.java

```diff
@@ -0,0 +1,168 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.AmbiguousSymbol;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.symbols.Usage;
+import org.sonar.plugins.python.api.tree.AssignmentStatement;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.InExpression;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.plugins.python.api.types.InferredType;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S5642")
+public class MembershipTestSupportCheck extends PythonSubscriptionCheck {
+  private static final String PRIMARY_MESSAGE = "Change the type of %s";
+  private static final String PRIMARY_MESSAGE_MULTILINE = "Change the type for the target expression of `in`";
+  private static final String KNOWN_TYPE_MESSAGE = "; type %s does not support membership protocol.";
+  private static final String UNKNOWN_TYPE_MESSAGE = "; the type does not support the membership protocol.";
+  private static final String SECONDARY_MESSAGE = "The result value of this expression does not support the membership protocol.";
+
+  // The ordering of membership protocol methods matters here (for extreme edge cases).
+  //
+  // For instance, a class that has __contains__ set to None but which defines __iter__ does not fulfill the membership protocol
+  // A class that defines __contains__ and has __iter__ set to None does fulfill the membership protocol
+  //
+  // Hence, we check for these special methods in the same order as the python interpreter does at runtime.
+  private static final List<String> MEMBERSHIP_PROTOCOL_ENABLING_METHODS = List.of("__contains__", "__iter__", "__getitem__");
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.IN, ctx -> checkInExpression(ctx, (InExpression) ctx.syntaxNode()));
+  }
+
+  private static void checkInExpression(SubscriptionContext ctx, InExpression inExpression) {
+    Expression rhs = inExpression.rightOperand();
+    InferredType rhsType = rhs.type();
+
+    if (canSupportMembershipProtocol(rhsType)) {
+      return;
+    }
+
+    PreciseIssue primaryLocation = addIssueOnInAndNotIn(ctx, inExpression, genPrimaryMessage(rhs, rhsType));
+    primaryLocation.secondary(inExpression.rightOperand(), SECONDARY_MESSAGE);
```

Comment:
Minor: I would inline `primaryLocation` variable here:
```suggestion
    addIssueOnInAndNotIn(ctx, inExpression, genPrimaryMessage(rhs, rhsType))
      .secondary(inExpression.rightOperand(), SECONDARY_MESSAGE);
```